pip error on Macbook pro (m1 pro)

Hi! I am working on Macbook pro with basemap, but I met some issues. When I try to pip install basemap, there is something wrong with this. But I have already brew install geos, what's wrong? Thanks!
![image](https://user-images.githubusercontent.com/84232793/162112051-89b615c5-4af5-459a-a14c-6a52762a9062.png)

I have installed geos 3.9.2 now from source https://libgeos.org/usage/download/
I Installed basemap from source via https://github.com/matplotlib/basemap instructions
Debugging setup.py is beyond my knowledge level. I know it's late there, please let me know what I can try.

```
$  ls -l $GEOS_DIR/include/geos_c.h
-rw-r--r--  1 rbradfor  staff  79397 Apr  7 17:53 /usr/local/include/geos_c.h
$ python -m pip install .
DEPRECATION: Configuring installation scheme with distutils config files is deprecated and will no longer work in the near future. If you are using a Homebrew or Linuxbrew Python, please see discussion at https://github.com/Homebrew/homebrew-core/issues/76621
Processing /Users/rbradfor/tmp/basemap/packages/basemap
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy<1.23,>=1.21 in /usr/local/lib/python3.9/site-packages (from basemap==1.3.2+dev) (1.22.2)
Requirement already satisfied: basemap-data<1.4,>=1.3.2 in /usr/local/lib/python3.9/site-packages (from basemap==1.3.2+dev) (1.3.2)
Requirement already satisfied: matplotlib<3.6,>=1.5 in /usr/local/lib/python3.9/site-packages (from basemap==1.3.2+dev) (3.5.1)
Collecting pyshp<2.2,>=1.2
  Using cached pyshp-2.1.3-py3-none-any.whl
Requirement already satisfied: pyproj<3.4.0,>=1.9.3 in /usr/local/lib/python3.9/site-packages (from basemap==1.3.2+dev) (3.3.0)
Requirement already satisfied: pillow>=6.2.0 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (9.0.1)
Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (1.3.2)
Requirement already satisfied: pyparsing>=2.2.1 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (3.0.7)
Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (21.3)
Requirement already satisfied: fonttools>=4.22.0 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (4.29.1)
Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (0.11.0)
Requirement already satisfied: python-dateutil>=2.7 in /usr/local/lib/python3.9/site-packages (from matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (2.8.2)
Requirement already satisfied: certifi in /usr/local/lib/python3.9/site-packages (from pyproj<3.4.0,>=1.9.3->basemap==1.3.2+dev) (2021.10.8)
Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.9/site-packages (from python-dateutil>=2.7->matplotlib<3.6,>=1.5->basemap==1.3.2+dev) (1.16.0)
Building wheels for collected packages: basemap
  Building wheel for basemap (pyproject.toml) ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/local/opt/python@3.9/bin/python3.9 /usr/local/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /var/folders/hp/l11pqpjn721d4cjpw1qjzl_c0000gn/T/tmp8iq42jtw
       cwd: /Users/rbradfor/tmp/basemap/packages/basemap
  Complete output (19 lines):
  setup.py:52: RuntimeWarning: Cannot find GEOS library and/or headers in standard locations ('/usr/local'). Please install the corresponding packages using your software management system or set the environment variable GEOS_DIR to point to the location where GEOS is installed (for example, if 'geos_c.h' is in '/usr/local/include' and 'libgeos_c' is in '/usr/local/lib', then you need to set GEOS_DIR to '/usr/local'
    warnings.warn(" ".join([
  running bdist_wheel
  running build
  running build_py
  running build_ext
  clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk -I/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/include -I/usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/include/python3.9 -c src/_geoslib.c -o build/temp.macosx-10.15-x86_64-3.9/src/_geoslib.o
  In file included from src/_geoslib.c:685:
  In file included from /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/include/numpy/arrayobject.h:5:
  In file included from /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/include/numpy/ndarrayobject.h:12:
  In file included from /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/include/numpy/ndarraytypes.h:1960:
  /usr/local/Cellar/python@3.9/3.9.10/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/numpy/core/include/numpy/npy_1_7_deprecated_api.h:17:2: warning: "Using deprecated NumPy API, disable it with "          "#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION" [-W#warnings]
  #warning "Using deprecated NumPy API, disable it with " \
   ^
  src/_geoslib.c:686:10: fatal error: 'geos_c.h' file not found
  #include "geos_c.h"
           ^~~~~~~~~~
  1 warning and 1 error generated.
  error: command '/usr/bin/clang' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for basemap
Failed to build basemap
ERROR: Could not build wheels for basemap, which is required to install pyproject.toml-based projects
WARNING: You are using pip version 21.3.1; however, version 22.0.4 is available.
You should consider upgrading via the '/usr/local/opt/python@3.9/bin/python3.9 -m pip install --upgrade pip' command.
```

--- a/packages/basemap/setup.py
+++ b/packages/basemap/setup.py
@@ -37,12 +37,14 @@ def get_geos_install_prefix():
         candidates = [os.path.expanduser("~/local"), os.path.expanduser("~"),
                       "/usr/local", "/usr", "/opt/local", "/opt", "/sw"]
 
+    # Prepare filename pattern to find the GEOS library.
+    extensions = {"win32": "dll", "cygwin": "dll", "darwin": "dylib"}
+    libext = extensions.get(sys.platform, "so*")
+    libname = "*geos_c*.{0}".format(libext)
+    libdirs = ["bin", "lib", "lib/x86_64-linux-gnu", "lib64"]
+
     for prefix in candidates:
         libfiles = []
-        libdirs = ["bin", "lib", "lib/x86_64-linux-gnu", "lib64"]
-        libext = "dll" if os.name == "nt" else "so"
-        libcode = "{0}geos_c".format("" if os.name == "nt" else "lib")
-        libname = "{0}*.{1}*".format(libcode, libext)
         for libdir in libdirs:
             libfiles.extend(glob.glob(os.path.join(prefix, libdir, libname)))
         hfile = os.path.join(prefix, "include", "geos_c.h")
@@ -111,7 +113,7 @@ if geos_install_prefix is not None:
         #   `distutils` bug (http://bugs.python.org/issue2437).
         library_dirs.append(os.path.join(geos_install_prefix, "bin"))
         runtime_library_dirs = []
-        dlls = glob.glob(os.path.join(geos_install_prefix, "*", "geos_c.dll"))
+        dlls = glob.glob(os.path.join(geos_install_prefix, "*", "*geos_c.dll"))
         if dlls:
             data_files.append(("../..", sorted(dlls)))
 

--- a/packages/basemap/setup.py
+++ b/packages/basemap/setup.py
@@ -119,7 +119,7 @@ if geos_install_prefix is not None:
         #   `distutils` bug (http://bugs.python.org/issue2437).
         library_dirs.append(os.path.join(geos_install_prefix, "bin"))
         runtime_library_dirs = []
-        dlls = glob.glob(os.path.join(geos_install_prefix, "*", "*geos_c.dll"))
+        dlls = glob.glob(os.path.join(geos_install_prefix, "*", "geos_c.dll"))
         if dlls:
             data_files.append(("../..", sorted(dlls)))
 

--- a/packages/basemap/setup.py
+++ b/packages/basemap/setup.py
@@ -51,15 +51,19 @@ def get_geos_install_prefix():
         if os.path.isfile(hfile) and libfiles:
             return prefix
 
-    warnings.warn(" ".join([
-        "Cannot find GEOS library and/or headers in standard locations",
-        "('{0}'). Please install the corresponding packages using your",
-        "software management system or set the environment variable",
-        "GEOS_DIR to point to the location where GEOS is installed",
-        "(for example, if 'geos_c.h' is in '/usr/local/include'",
-        "and 'libgeos_c' is in '/usr/local/lib', then you need to",
-        "set GEOS_DIR to '/usr/local'",
-    ]).format("', '".join(candidates)), RuntimeWarning)
+    # At this point, the GEOS library was not found, so we throw a warning if
+    # the user is trying to build the library.
+    build_cmds = ("bdist_wheel", "build", "install")
+    if any(cmd in sys.argv[1:] for cmd in build_cmds):
+        warnings.warn(" ".join([
+            "Cannot find GEOS library and/or headers in standard locations",
+            "('{0}'). Please install the corresponding packages using your",
+            "software management system or set the environment variable",
+            "GEOS_DIR to point to the location where GEOS is installed",
+            "(for example, if 'geos_c.h' is in '/usr/local/include'",
+            "and 'libgeos_c' is in '/usr/local/lib', then you need to",
+            "set GEOS_DIR to '/usr/local'",
+        ]).format("', '".join(candidates)), RuntimeWarning)
     return None
 
 
@@ -96,7 +100,9 @@ else:
         import numpy
         include_dirs.append(numpy.get_include())
     except ImportError as err:
-        warnings.warn("unable to locate NumPy headers", RuntimeWarning)
+        build_cmds = ("bdist_wheel", "build", "install")
+        if any(cmd in sys.argv[1:] for cmd in build_cmds):
+            warnings.warn("unable to locate NumPy headers", RuntimeWarning)
 
 # Define GEOS include, library and runtime dirs.
 geos_install_prefix = get_geos_install_prefix()

--- a/packages/basemap/utils/GeosLibrary.py
+++ b/packages/basemap/utils/GeosLibrary.py
@@ -147,12 +147,24 @@ class GeosLibrary(object):
             for line in lines:
                 fd.write(line.replace(oldtext, newtext).encode())
 
-        # The SVN revision file is not created on the fly before 3.6.0.
-        svn_hfile = os.path.join(zipfold, "geos_svn_revision.h")
-        if self.version_tuple < (3, 6, 0) and not os.path.exists(svn_hfile):
-            with io.open(svn_hfile, "wb") as fd:
-                text = "#define GEOS_SVN_REVISION 0"
-                fd.write(text.encode())
+        # Apply specific patches for GEOS < 3.6.0.
+        if self.version_tuple < (3, 6, 0):
+            # The SVN revision file is not created on the fly before 3.6.0.
+            svn_hfile = os.path.join(zipfold, "geos_svn_revision.h")
+            if not os.path.exists(svn_hfile):
+                with io.open(svn_hfile, "wb") as fd:
+                    text = "#define GEOS_SVN_REVISION 0"
+                    fd.write(text.encode())
+            # Reduce warnings when compiling with `nmake` on Windows.
+            cmakefile = os.path.join(zipfold, "CMakeLists.txt")
+            if os.path.exists(cmakefile):
+                with io.open(cmakefile, "r", encoding="utf-8") as fd:
+                    lines = fd.readlines()
+                with io.open(cmakefile, "wb") as fd:
+                    oldtext = 'string(REGEX REPLACE "/W[0-9]" "/W4"'
+                    newtext = oldtext.replace("W4", "W1")
+                    for line in lines:
+                        fd.write(line.replace(oldtext, newtext).encode())
 
     def build(self, installdir=None, njobs=1):
         """Build and install GEOS from source."""
@@ -191,7 +203,7 @@ class GeosLibrary(object):
             build_opts.extend([
                 "--",
                 "WIN64={0}".format("YES" if win64 else "NO"),
-                "BUILD_BATCH={0}".format("YES" if njobs > 1 else "NO")
+                "BUILD_BATCH={0}".format("YES" if njobs > 1 else "NO"),
             ])
         else:
             build_opts = ["-j", "{0:d}".format(njobs)] + build_opts

